finetune/datasets/hugging_face/hugging_face_subset_loader.py

```diff
@@ -0,0 +1,101 @@
+        sequence_length,
+        page_row_offsets: typing.List[int],
+        page_row_count: int,
+        tokenizer: AutoTokenizer,
+    ):
+        self.batch_size = batch_size
+        self.sequence_length = sequence_length
+        self.tokenizer = tokenizer
+        self.base_url = "https://datasets-server.huggingface.co/rows"
+        self.params = {
+            "dataset": dataset_name,
+            "config": "default",
+            "split": "train",
+        }
+        self.page_row_offsets = page_row_offsets
+        self.page_row_count = page_row_count
+        self.buffer = []
+        self.retry_limit = 10  # Number of retries
+        self.retry_delay = 5  # Seconds to wait between retries
+
+        for page_row_offset in self.page_row_offsets:
+            self.fetch_data_for_page(page_row_offset)
+
+    def fetch_data_for_page(self, page_row_offset):
+        self.params["offset"] = page_row_offset
+        self.params["limit"] = self.page_row_count
+        attempt = 0
+        while attempt < self.retry_limit:
+            try:
+                response = requests.get(self.base_url, params=self.params)
+                response.raise_for_status()  # This will raise an HTTPError if the HTTP request returned an unsuccessful status code
+                for row in response.json()["rows"]:
+                    content = row["row"]["content"]
+                    # Truncate/Pad the content to the desired sequence length. This avoid context breaks.
+                    self.buffer += self.tokenizer(
+                        content,
+                        truncation=True,
+                        padding="max_length",
+                        max_length=self.sequence_length,
+                    )["input_ids"]
+                break  # If the request was successful, break out of the retry loop
+            except requests.exceptions.RequestException as e:
+                attempt += 1
+                bt.logging.warning(
+                    f"Failed to fetch data, retrying. Attempt {attempt}/{self.retry_limit}"
+                )
+                if attempt < self.retry_limit:
+                    time.sleep(self.retry_delay)  # Wait before the next retry
+                else:
+                    bt.logging.error(
+                        "Maximum retry limit reached. Unable to fetch data."
+                    )
+                    raise
+
+    def __iter__(self):
+        while len(self.buffer) >= self.sequence_length * self.batch_size:
+            batch = []
+            for _ in range(self.batch_size):
+                batch.append(torch.tensor(self.buffer[: self.sequence_length]))
+                self.buffer = self.buffer[self.sequence_length :]
+            yield torch.stack(batch)
+
+    def __next__(self):
+        batch = []
+        for _ in range(self.batch_size):
+            batch.append(torch.tensor(self.buffer[: self.sequence_length]))
+            self.buffer = self.buffer[self.sequence_length :]
```

Comment:
Buffer could be changed to just a list of batches now given that each row is just one batch.

If doing so we could also avoid padding and just use truncation as well. Unsure if we will want to always have batches be exactly sequence length or not though in which case keeping the padding makes sense.